crds/parse.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+
+	yamlExtras "github.com/ghodss/yaml"
+	"gopkg.in/yaml.v2"
+)
+
+// ResourceDefinition contains the minimum ammount of information of a
+// Kubernetes CRD to generate appropiate constructors/interfaces/classes.
+type ResourceDefinition struct {
+	json        string
+	kind        string
+	versionName string
+	group       string
+	fields      []Field
+}
+
+func (r ResourceDefinition) apiVersion() string {
+	return fmt.Sprintf("%s/%s", r.group, r.versionName)
+}
+
+func (r ResourceDefinition) argsName() string {
+	return r.kind + "Args"
+}
+
+func (r ResourceDefinition) specArgsName() string {
+	return r.kind + "SpecArgs"
+}
+
+func (r ResourceDefinition) definitionName() string {
+	return r.kind + "Definition"
+}
+
+type Field struct {
+	name     string
+	openType string
+	required bool
+}
+
+// This is a utility function meant to linearly scan and check if a string is
+// contained within a string slice. Since the CRD spec provides us with a list
+// of required fields, we need to call this function for every field we process
+// to check if it is required. This makes it O(n^2) time then, so if it turns
+// out that some CRD required field lists are extremely large, then I'll
+// implement some set class to get it down to O(n).
+func contains(list []string, itemToFind string) bool {
+	for _, item := range list {
+		if item == itemToFind {
+			return true
+		}
+	}
+	return false
+}
+
+// NewResourceDefinition returns a ResourceDefinition by parsing out all
+// required information from a NestedMap.
+func NewResourceDefinition(yamlFile []byte) ResourceDefinition {
```

Comment:
I'd recommend taking a look at this code: https://github.com/pulumi/pulumi-kubernetes/blob/b493881669be46a881a0b4ca96d6d5656bb42d76/provider/pkg/provider/invoke_decode_yaml.go#L32-L41

This method uses the Kubernetes packages to decode the YAML into the `Unstructured` type. This type contains some convenience methods for working with k8s resources, but mostly relies on `map[string]interface{}`. Unlike the built in k8s resources, we don't have types defined for CustomResources, so we're stuck with the generic map type.